docs/amdgpu_kernel_optimization_guide.md

```diff
@@ -293,3 +293,108 @@ forms a *clause* that translates to a single data fabric transaction.
+
+The cons are
+1. It only supports limited patterns
+2. Similar to `ds_bpermute`, `s_waitcnt` is still required to wait for `dest` VGPR.
+
+The instruction syntax:
+```nasm
+ds_swizzle_b32 dest, src offset:ds_pattern
+```
+
+### Data-Parallel Primitives, DPP
+
+DPP is a 32-bit instruction modifier appended to the normal VALU instructions. It
+allows VALU instructions to access data in neighboring lanes directly, which means
+it doesn't need LDS hardware anymore, so no `s_waitcnt` instructions are required.
+
+Unfortunately, it also supported limited patterns like `ds_swizzle`. And there are
+some instructions that can't be modified by DPP.
+
+The instruction syntax:
+```nasm
+; Normal inst
+v_add_f32
+
+; Inst modified by dpp
+v_add_f32_dpp
+```
+
+It's worth mentioning that DPP has different names and syntaxes on different architectures.
+- CDNA: DPP
+- RDNA: DPP8/DPP16
+
+For details, please check [MI300 ISA Reference Guide](https://www.amd.com/content/dam/amd/en/documents/instinct-tech-docs/instruction-set-architectures/amd-instinct-mi300-cdna3-instruction-set-architecture.pdf) and
+[RDNA3 ISA Reference Guide](https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/rdna3-shader-instruction-set-architecture-feb-2023_0.pdf)
+
+### How to use them in MLIR
+
+Each instruction has its corresponding Op in MLIR(except `ds_permute`):
+- `ds_bpermute`: `rocdl.ds_bpermute`
+- `ds_swizzle`: `rocdl.ds_swizzle`
+- DPP: `rocdl.update.dpp`, `amdgpu.dpp`(a slim wrapper of `rocdl.update.dpp` with more comprehensive user interface)
+
+The first 2 are straightforward, while DPP follows a different fashion.
+
+Since DPP is an instruction modifier instead of an instruction itself, there are
+tremendous number of combinations of VALU instructions and DPP. To solve that, `rocdl.update.dpp`
+and `amdgpu.dpp` are designed to be a wrapper of `v_mov_b32_dpp` instruction. And it depends
+on LLVM compiler to fuse it with the subsequent VALU instruction **with best efforts**.
+
+For example, `v_mov_b32_dpp` + `v_add_f32_e32` might be fused into `v_add_f32_dpp`.
+
+There are plenty of constraints stopping an instruction from being merged.
+For example, if either the `bank_mask` or the `row_mask` is not `0xf`, it can't be fused.
+You can check the [GCNDPPCombine::combineDPPMov](https://github.com/llvm/llvm-project/blob/llvmorg-19.1.3/llvm/lib/Target/AMDGPU/GCNDPPCombine.cpp#L522) function to see how it works.
+
+### Comparison
+
+To summarize, there's no free lunch. You can't have both performance and generality
+at the same time.
+
+The performance ranking is like:
+
+DPP > `ds_swizzle` >= `ds_permute`/`ds_bpermute`
+
+while the generality ranking is the opposite:
+
+DPP < `ds_swizzle` < `ds_permute`/`ds_bpermute`
+
+This table summarized some useful facts. The latency data is collected from an experiment
+on Fused Softmax kernel with [rocprofv2](https://github.com/ROCm/rocprofiler?tab=readme-ov-file#plugin-support):
```

Comment:
Also mention the hardware. I assume this is on mi300?

Reply:
Yes